packages/flutter_tools/lib/src/base/net.dart

```diff
@@ -12,31 +13,46 @@ const int kNetworkProblemExitCode = 50;
 
 /// Download a file from the given URL and return the bytes.
 Future<List<int>> fetchUrl(Uri url) async {
-  printTrace('Downloading $url.');
+  int attempts = 0;
+  int duration = 1;
+  while (true) {
+    attempts += 1;
+    final List<int> result = await _attempt(url);
+    if (result != null)
+      return result;
+    printStatus('Download failed -- attempting retry $attempts in $duration second${ duration == 1 ? "" : "s"}...');
+    await new Future<Null>.delayed(new Duration(seconds: duration));
+    if (duration < 64)
+      duration *= 2;
+  }
+}
 
-  final HttpClient httpClient = new HttpClient();
+Future<List<int>> _attempt(Uri url) async {
+  printTrace('Downloading: $url');
+  final HttpClient httpClient = context[HttpClient] ?? new HttpClient();
```

Comment:
It's a little weird that we'll instantiate a new one in regular operation but reuse on in tests.  How about sticking a HttpCientFactory in the context instead, defaulting to one that just instantiates `new HttpClient()`?

```
typedef HttpClient HttpClientFactory();
```

...then you can even make the test code reuse the same instance in its factory, but at least the lib code looks consistent in how it gets an HttpClient.

Reply:
I could make it a cached static, too, so it's reused in normal operation (though maybe that's not a good thing? Maybe we want to forget cookies for instance?).

Reply:
Yeah, reusing the instance seems like a recipe for a problem down the road.

Reply:
`HttpClientFactory: () => () => new MockHttpClient(500)`